datalad/distribution/uninstall.py

```diff
@@ -12,72 +12,91 @@
+from datalad.interface.common_opts import recursion_flag
+from datalad.interface.utils import handle_dirty_dataset
+from datalad.utils import rmtree
+from datalad.utils import getpwd
 
 lgr = logging.getLogger('datalad.distribution.uninstall')
 
 
-def _move_gitdir(ds, relativepath):
-    """Move .git directory of submodule `relativepath` into .git/modules of `ds`
+def _uninstall(ds, paths, check, remove_history, remove_data, remove_handles,
+               recursive):
+    # all input paths are assumed to belong to the given dataset,
+    # and in particular not to any subdataset
 
-    After moving, this will create a .git file in the submodule instead.
+    # recode paths to be relative to the root of the dataset
+    paths = [relpath(p, start=ds.path) for p in paths]
 
-    Parameters
-    ds: Dataset
-    relativepath: str
+    if os.curdir in paths:
+        # we can take a shortcut if the entire thing goes away
+        lgr.debug('uninstall entire content in {}'.format(ds))
+        paths = [os.curdir]
 
-    """
-
-    # relativepath is expected to point to a submodule of ds:
-    assert(relativepath in ds.get_subdatasets())
-
-    path = opj(ds.path, relativepath)
-    src_dotgit = get_git_dir(path)
-
-    # expect the actual git dir to be within submodule:
-    assert(src_dotgit == '.git')
-
-    # get actual path to the .git to be moved:
-    src_git_dir = opj(path, src_dotgit)
+    results = []
 
-    # move .git to superrepo's .git/modules, remove .git, create
-    # .git-file/symlink
+    # at this point we can only have two things:
+    # 1. files and directories in the current dataset
+    # 2. the entire dataset
 
-    ds_git_dir = get_git_dir(ds.path)
-    moved_git_dir = opj(ds.path, ds_git_dir, "modules", relativepath)
-    assure_dir(moved_git_dir)
-    from os import rename, listdir, rmdir
-    for dot_git_entry in listdir(src_git_dir):
-        rename(opj(src_git_dir, dot_git_entry),
-               opj(moved_git_dir, dot_git_entry))
-    assert not listdir(src_git_dir)
-    rmdir(src_git_dir)
-
-    # write .git file
-    # Note: Annex would need a symlink instead to not break sub-submodules. But
-    #       as of version 2.7.0 git itself doesn't deal correctly with
-    #       .git-symlinks when deinitializing a submodule. Instead, it will
-    #       still complain about a .git dir present in the submodule.
-    #       Since by now this is about deinitializing only, we go for the .git
-    #       file and don't care for possible sub-annexes.
-    target_path = relpath(moved_git_dir, start=path)
-    with open(opj(path, ".git"), "w") as f:
-        f.write("gitdir: {moved}\n".format(moved=target_path))
+    # alway drop data first
+    if remove_data:
+        if hasattr(ds.repo, 'drop'):
+            opts = ['--force'] if not check else []
+            results.extend(ds.repo.drop(paths, options=opts))
+        else:
+            lgr.warning("cannot uninstall data only, not an annex dataset")
```

Comment:
it would be good to provide location - since otherwise there might be just a flood of not so informative warnings in case of recursive